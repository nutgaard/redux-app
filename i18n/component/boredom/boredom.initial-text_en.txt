Click the button to procastinate
