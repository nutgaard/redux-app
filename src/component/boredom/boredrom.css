.boredom {
  padding: 1rem;
}

.boredom__button {
  margin-bottom: 0.75rem;
}
