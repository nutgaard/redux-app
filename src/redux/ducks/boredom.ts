import { Action } from 'redux';
import { ThunkAction } from 'redux-thunk';
import { assertNever } from '../utils';
import { AppState } from '../appstate';

enum TypeKeys {
    FIND_NEW_REQUEST = 'BOREDOM/FIND_NEW_REQUEST',
    FIND_NEW_OK = 'BOREDOM/FIND_NEW_OK',
    FIND_NEW_ERROR = 'BOREDOM/FIND_NEW_ERROR'
}

interface FindNewRequest extends Action<TypeKeys.FIND_NEW_REQUEST> {
}

interface FindNewError extends Action<TypeKeys.FIND_NEW_ERROR> {
    error: Error;
}

interface FindNewOk extends Action<TypeKeys.FIND_NEW_OK> {
    activity: string;
}

type Actions = FindNewRequest | FindNewError | FindNewOk;

interface OkState {
    type: TypeKeys.FIND_NEW_OK;
    activity: string;
}

interface ErrorState {
    type: TypeKeys.FIND_NEW_ERROR;
    error: Error;
}

interface RequestingState {
    type: TypeKeys.FIND_NEW_REQUEST;
    start: Date
}

export type BoredomState = OkState | ErrorState | RequestingState;

const initialState: BoredomState = {
    type: TypeKeys.FIND_NEW_OK,
    activity: 'Click the button to procastinate'
};

export default function reducer(state: BoredomState = initialState, action: Actions): BoredomState {
    switch (action.type) {
        case TypeKeys.FIND_NEW_REQUEST: {
            return {
                type: action.type,
                start: new Date()
            };
        }
        case TypeKeys.FIND_NEW_ERROR: {
            return {
                type: action.type,
                error: action.error
            };
        }
        case TypeKeys.FIND_NEW_OK: {
            return {
                type: action.type,
                activity: action.activity
            };
        }
        default: {
            assertNever(action);
            return state;
        }
    }
};

interface BoredomData {
    activity: string;
}

export function findNew(): ThunkAction<Promise<BoredomData>, AppState, {}, Actions> {
    return (dispatch) => {
        dispatch({ type: TypeKeys.FIND_NEW_REQUEST });

        return fetch(`https://www.boredapi.com/api/activity`)
            .then((resp) => {
                if (!resp.ok) {
                    throw new Error(`Status: ${resp.status}`);
                }
                return resp.json();
            })
            .then((json) => {
                dispatch({ type: TypeKeys.FIND_NEW_OK, activity: json.activity });
                return json;
            }).catch((error: Error) => {
                dispatch({ type: TypeKeys.FIND_NEW_ERROR, error });
                return error;
            })
    }

}

export function selectBoredomStatus(state: AppState): string {
    const boredom = state.boredom;
    switch (boredom.type) {
        case TypeKeys.FIND_NEW_REQUEST:
            return `Requesting`;
        case TypeKeys.FIND_NEW_OK:
            return boredom.activity;
        case TypeKeys.FIND_NEW_ERROR:
            return boredom.error.message;
        default:
            assertNever(boredom);
            return "Unknown...";
    }
}
