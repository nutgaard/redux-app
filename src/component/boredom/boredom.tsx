import * as React from 'react';
import { bindActionCreators, Dispatch } from 'redux';
import { connect } from 'react-redux';
import { FormattedMessage } from 'react-intl';
import { findNew, selectBoredomStatus } from '../../redux/ducks/boredom';
import { AppState } from '../../redux/appstate';
import Intl from './intl';
import './boredrom.css';

interface StateProps {
    boredomResponse?: string;
}
interface DispatchProps {
    findNew: typeof findNew
}
type Props = StateProps & DispatchProps;

function Boredom(props: Props) {
    let text: string | JSX.Element = props.boredomResponse!;
    if (!text) {
        text = <FormattedMessage id={Intl.boredomInitialText}/>;
    }

    return (
        <div className="boredom">
            <button className="boredom__button" onClick={props.findNew}>
                <FormattedMessage id={Intl.boredomFindNext}/>
            </button>
            <p>{text}</p>
        </div>
    );
}

const mapStateToProps = (state: AppState): StateProps => ({
    boredomResponse: selectBoredomStatus(state)
});
const mapDispatchToProps = (dispatch: Dispatch): DispatchProps => ({
    ...bindActionCreators({
        findNew
    }, dispatch)
});

export default connect(mapStateToProps, mapDispatchToProps)(Boredom);
