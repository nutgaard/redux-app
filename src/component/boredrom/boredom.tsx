import * as React from 'react';
import { bindActionCreators, Dispatch } from 'redux';
import { connect } from 'react-redux';
import { findNew, selectBoredomStatus } from '../../redux/ducks/boredom';
import { AppState } from '../../redux/appstate';
import './boredrom.css';

interface StateProps {
    boredomResponse: string;
}
interface DispatchProps {
    findNew: typeof findNew
}
type Props = StateProps & DispatchProps;

function Boredom(props: Props) {
    return (
        <div className="boredom">
            <button className="boredom__button" onClick={props.findNew}>Find activity</button>
            <p>{props.boredomResponse}</p>
        </div>
    );
}

const mapStateToProps = (state: AppState): StateProps => ({
    boredomResponse: selectBoredomStatus(state)
});
const mapDispatchToProps = (dispatch: Dispatch): DispatchProps => ({
    ...bindActionCreators({
        findNew
    }, dispatch)
});

export default connect(mapStateToProps, mapDispatchToProps)(Boredom);
