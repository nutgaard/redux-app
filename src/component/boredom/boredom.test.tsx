import * as React from 'react';
import { mount } from 'enzyme';
import { IntlProvider } from 'react-intl';
import { Boredom } from './boredom';
import messages from '../../bundle';
import { ReactNode } from 'react';

function mountWithIntl(node: ReactNode) {
    return mount(
        <IntlProvider locale="en" messages={messages.en}>
            {node}
        </IntlProvider>
    ).get(0);
}

describe('boredom component', () => {
    it('boredom snapshot', () => {
        const findNew = jest.fn();
        const wrapper = mountWithIntl(<Boredom findNew={findNew} />);

        expect(wrapper).toMatchSnapshot();
    });
});
