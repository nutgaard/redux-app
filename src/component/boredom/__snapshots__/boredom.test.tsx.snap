// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`boredom component boredom snapshot 1`] = `
<IntlProvider
  locale="en"
  messages={
    Object {
      "boredom.find-next": "Find activity",
      "boredom.initial-text": "Click the button to procastinate",
      "counter.decrement": "-",
      "counter.increment": "+",
      "counter.step-ordinal": "{steps, selectordinal,
    one {#st}
    two {#nd}
    few {#rd}
    other {#th}
}",
      "counter.step-plural": "{steps, plural,
    =0 {no steps}
    one {1 step}
    other {{steps} steps}
}",
      "counter.stepsize": "Stepsize:",
      "hero.header": "Welcome to React",
      "hero.intro": "To get started, edit <code>src/app.tsx</code> and save to reload.",
      "summary.boredom-text": "Current activity:",
      "summary.counter-stepsize": "Stepsize:",
      "summary.counter-value": "Counter value:",
      "summary.fibonacci-text": "Fibonacci:",
    }
  }
>
  <Boredom
    findNew={[MockFunction]}
  >
    <div
      className="boredom"
    >
      <button
        className="boredom__button"
        onClick={[MockFunction]}
      >
        <FormattedMessage
          id="boredom.find-next"
          values={Object {}}
        >
          <span>
            Find activity
          </span>
        </FormattedMessage>
      </button>
      <p>
        <FormattedMessage
          id="boredom.initial-text"
          values={Object {}}
        >
          <span>
            Click the button to procastinate
          </span>
        </FormattedMessage>
      </p>
    </div>
  </Boredom>
</IntlProvider>
`;
